src/diffusers/quantizers/gguf/utils.py

```diff
@@ -0,0 +1,384 @@
+# Copyright 2024 The HuggingFace Team and City96. All rights reserved.
+# #
+# # Licensed under the Apache License, Version 2.0 (the "License");
+# # you may not use this file except in compliance with the License.
+# # You may obtain a copy of the License at
+# #
+# #     http://www.apache.org/licenses/LICENSE-2.0
+# #
+# # Unless required by applicable law or agreed to in writing, software
+# # distributed under the License is distributed on an "AS IS" BASIS,
+# # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# # See the License for the specific language governing permissions and
+# # limitations under the License.
+
+
+import gguf
+import torch
+import torch.nn as nn
+
+
+def _replace_with_gguf_linear(model, compute_dtype, state_dict, prefix=""):
+    def _should_convert_to_gguf(module, state_dict, prefix):
+        weight_key = prefix + "weight"
+        return weight_key in state_dict and isinstance(state_dict[weight_key], GGUFParameter)
+
+    has_children = list(model.children())
+    if not has_children:
+        return
+
+    for name, module in model.named_children():
+        module_prefix = prefix + name + "."
+        _replace_with_gguf_linear(module, compute_dtype, state_dict, module_prefix)
+
+        if isinstance(module, nn.Linear) and _should_convert_to_gguf(module, state_dict, module_prefix):
+            model._modules[name] = GGUFLinear(
+                module.in_features,
+                module.out_features,
+                module.bias is not None,
+                compute_dtype=compute_dtype,
+            )
+            model._modules[name].source_cls = type(module)
+            # Force requires grad to False to avoid unexpected errors
+            model._modules[name].requires_grad_(False)
+
+    return model
+
+
+# dequantize operations based on torch ports of GGUF dequantize_functions
+# from City96
+# more info: https://github.com/city96/ComfyUI-GGUF/blob/main/dequant.py
+
+
+QK_K = 256
+K_SCALE_SIZE = 12
+
+
+def to_uint32(x):
+    x = x.view(torch.uint8).to(torch.int32)
+    return (x[:, 0] | x[:, 1] << 8 | x[:, 2] << 16 | x[:, 3] << 24).unsqueeze(1)
+
+
+def split_block_dims(blocks, *args):
+    n_max = blocks.shape[1]
+    dims = list(args) + [n_max - sum(args)]
+    return torch.split(blocks, dims, dim=1)
+
+
+def get_scale_min(scales):
+    n_blocks = scales.shape[0]
+    scales = scales.view(torch.uint8)
+    scales = scales.reshape((n_blocks, 3, 4))
+
+    d, m, m_d = torch.split(scales, scales.shape[-2] // 3, dim=-2)
+
+    sc = torch.cat([d & 0x3F, (m_d & 0x0F) | ((d >> 2) & 0x30)], dim=-1)
+    min = torch.cat([m & 0x3F, (m_d >> 4) | ((m >> 2) & 0x30)], dim=-1)
+
+    return (sc.reshape((n_blocks, 8)), min.reshape((n_blocks, 8)))
+
+
```

Comment:
gguf library have a dequantize function, can't be make use of it ? 

`def dequantize(data: np.ndarray, qtype: GGMLQuantizationType) -> np.ndarray:`

Reply:
Great to see you again on `diffusers`, @Isotr0py!

> BTW, if we want a runtime optimized dequantize function, we might need a torch compatible GGUF kernel, just like what we do in vLLM...

That is a great point. Perhaps we could ship the first iteration without and make use of those kernels (iff possible) in a follow-up PR? WDYT?

Reply:
Agreed, we can port the GGUF kernel in a follow-up PR later, because I'm going to update the GGUF kernel implementation in vLLM since it's pretty out of date. (Although I haven't had bandwidth to start it. 😅)